pkg/store/labelpb/label.go

```diff
@@ -0,0 +1,280 @@
+// Package containing Zero Copy Labels adapter.
+
+package labelpb
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"unsafe"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+func noAllocString(buf []byte) string {
+	return *((*string)(unsafe.Pointer(&buf)))
+}
+
+// LabelsFromPromLabels converts Prometheus labels to slice of labelpb.Label in type unsafe manner.
+// It reuses the same memory. Caller should abort using passed labels.Labels.
+func LabelsFromPromLabels(lset labels.Labels) []Label {
+	return *(*[]Label)(unsafe.Pointer(&lset))
+}
+
+// LabelsToPromLabels convert slice of labelpb.Label to Prometheus labels in type unsafe manner.
+// It reuses the same memory. Caller should abort using passed []Label.
+func LabelsToPromLabels(lset []Label) labels.Labels {
+	return *(*labels.Labels)(unsafe.Pointer(&lset))
+}
+
+// LabelSetsToPromLabelSets converts slice of labelpb.LabelSet to slice of Prometheus labels.
+func LabelSetsToPromLabelSets(lss ...LabelSet) []labels.Labels {
+	res := make([]labels.Labels, 0, len(lss))
+	for _, ls := range lss {
+		res = append(res, ls.PromLabels())
+	}
+	return res
+}
+
+// Label is a labels.Label that can be marshalled to/from protobuf reusing the same
+// memory address for string bytes.
+type Label labels.Label
+
+func (m *Label) Marshal() (data []byte, err error) {
+	size := m.Size()
+	data = make([]byte, size)
+	n, err := m.MarshalToSizedBuffer(data[:size])
+	if err != nil {
+		return nil, err
+	}
+	return data[:n], nil
+}
+
+func (m *Label) MarshalTo(data []byte) (int, error) {
+	size := m.Size()
+	return m.MarshalToSizedBuffer(data[:size])
+}
+
+func (m *Label) MarshalToSizedBuffer(data []byte) (int, error) {
+	i := len(data)
+	_ = i
+	var l int
+	_ = l
+	if len(m.Value) > 0 {
+		i -= len(m.Value)
+		copy(data[i:], m.Value)
+		i = encodeVarintTypes(data, i, uint64(len(m.Value)))
+		i--
+		data[i] = 0x12
+	}
+	if len(m.Name) > 0 {
+		i -= len(m.Name)
+		copy(data[i:], m.Name)
+		i = encodeVarintTypes(data, i, uint64(len(m.Name)))
+		i--
+		data[i] = 0xa
+	}
+	return len(data) - i, nil
+}
+
+func (m *Label) Unmarshal(data []byte) error {
+	l := len(data)
+	iNdEx := 0
```

Comment:
🤔 A new kind of case convention?

Reply:
Copied from protobuf marshal - so we left that this way.